lib/mongo/cluster/topology/server_description_list.rb

```diff
@@ -0,0 +1,56 @@
+# Copyright (C) 2018 MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module Mongo
+  class Cluster
+    module Topology
+      # A list of server descriptions. Stored in a topology.
+      #
+      # This class adds methods for retrieving server descriptions
+      # by their address.
+      #
+      # @since 2.7.0
+      class ServerDescriptionList < Array
```

Comment:
Is there a reason we don't use use a hash of addresses to server descriptions? Most other SDAM implementations I've seen use this pattern

Reply:
This implementation follows the spec tests which use a list of server descriptions. Which drivers use hashes?

Reply:
Python, for one. See [here](https://github.com/mongodb/mongo-python-driver/blob/341b2b5964b246fe7419a812e31b84e023c7c030/pymongo/topology_description.py#L44-L53) where they initialize the list of server descriptions in the topology description to a dictionary and [here](https://github.com/mongodb/mongo-python-driver/blob/341b2b5964b246fe7419a812e31b84e023c7c030/pymongo/topology.py#L110) where they initialize the list of servers in the topology to a dictionary.

More generally, it just seems like a more intuitive way to store things to me. Given that we need to be able to look up servers by the address and never care about the order of them in the container, a hash seems like the most logical choice. 

The spec does use an array [here](https://github.com/mongodb/specifications/blob/97fd1f26c3fa7506353f7e310063822b958a53b0/source/server-discovery-and-monitoring/tests/rs/new_primary.yml#L8) for the isMaster responses, but that's specifically because they need to be applied to the cluster in a certain order. (I'm honestly not sure why they make each element of that array an array as well; I think an object with the address as a key would make more sense). However, when describing the expected state of the topology after applying the isMaster responses, the spec does use an object, not an array, as shown [here](https://github.com/mongodb/specifications/blob/97fd1f26c3fa7506353f7e310063822b958a53b0/source/server-discovery-and-monitoring/tests/rs/new_primary.yml#L21-L36).

Reply:
Okay I will change it to be a hash.